Excel Import/Export/Report

kittiu opened the pull request:
The module provide pre-built functions and wizards for developer to build excel import / export / report with ease.

Without having to code to create excel file, developer do,

- Create menu, action, wizard, model, view a normal Odoo development.
- Design excel template using standard excel application, e.g., colors, fonts, formulas, etc.
- Instruct how the data will be located in excel with simple dictionary instruction or from Odoo UI.
- Odoo will combine instruction with excel template, and result in final excel file.

![00420](https://user-images.githubusercontent.com/1973598/53646905-f110f400-3c6e-11e9-9331-0a3a63cff2b5.png)

![00421](https://user-images.githubusercontent.com/1973598/53646918-f40be480-3c6e-11e9-95be-294c9d5834d5.png)

Install excel_import_export_sample and start testing it.

kittiu committed:
[ADD] v12 excel_import_export

kittiu commented:
@elicoidal @bealdav @NL66278 I made somewhat improvement on this module, also the readme. I would hope for this module to be the tools that ease import/export/report with excel development.
If you have chance to test it, feel free to discuss.

NL66278 commented:
@kittiu Please add the required python libraries to requirements.txt in the root of the repository. And you can limit the amount of commits by squashing them. If commiting regularly, you can use git commit --amend to add your work to a previous commit or use --fixup on your commit.

elicoidal commented on excel_import_export/__manifest__.py:
```diff
@@ -0,0 +1,23 @@
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html)
+
+{'name': 'Excel Import/Export',
+ 'summary': 'Base module for easy way to develop excel import/export',
```
```suggestion
 'summary': 'Base module for easy way to develop Excel import/export',
```

elicoidal commented on excel_import_export/models/xlsx_template.py:
```diff
@@ -0,0 +1,448 @@
+    )
+    input_instruction = fields.Text(
+        string='Instruction (Input)',
+        help="This is used to construct instruction in tab Imort/Export",
```
```suggestion
        help="This is used to construct instruction in tab Import/Export",
```

elicoidal commented on excel_import_export/readme/DESCRIPTION.rst:
```diff
@@ -0,0 +1,8 @@
+
+- Create menu, action, wizard, model, view a normal Odoo development.
+- Design excel template using standard excel application, e.g., colors, fonts, formulas, etc.
+- Instruct how the data will be located in excel with simple dictionary instruction or from Odoo UI.
```
```suggestion
- Instruct how the data will be located in Excel with simple dictionary instruction or from Odoo UI.
```

elicoidal commented on excel_import_export/readme/DESCRIPTION.rst:
```diff
@@ -0,0 +1,8 @@
+Without having to code to create excel file, developer do,
+
+- Create menu, action, wizard, model, view a normal Odoo development.
+- Design excel template using standard excel application, e.g., colors, fonts, formulas, etc.
```
```suggestion
- Design excel template using standard Excel application, e.g., colors, fonts, formulas, etc.
```

elicoidal commented on excel_import_export/security/ir.model.access.csv:
```diff
@@ -0,0 +1,4 @@
+"id","name","model_id:id","group_id:id","perm_read","perm_write","perm_create","perm_unlink"
+xlsx_template_user,xlsx_template_user,model_xlsx_template,,1,1,1,1
```
should not you have 2 levels? manager to make the set up and users to be able to import/export?

kittiu replied:
I think we now aim for developer to use first. But if for some import/export action, I think developer can add groups security as they wish (per normal way).
But if in practical, there is need, we can add in next iteration.

elicoidal commented on excel_import_export/views/xlsx_report.xml:
```diff
@@ -0,0 +1,49 @@
+                    <field name="name" invisible="1"/>
+                    <field name="choose_template" invisible="1"/>
+                    <div states="choose">
+                        <label string="Choose Template: " for="template_id" attrs="{'invisible': [('choose_template', '=', False)]}"/>
```
please improve readability with one attribute per line

elicoidal commented on excel_import_export_sample/__init__.py:
```diff
@@ -0,0 +1,5 @@
+# Copyright 2019 Ecosoft Co., Ltd (http://ecosoft.co.th/)
```
module name suggestion excel_import_export_demo

kittiu replied:
OK

elicoidal commented on excel_import_export/views/xlsx_template_view.xml:
```diff
@@ -0,0 +1,230 @@
+                                <h3>Sample Input Instruction as Dictionary</h3>
+                                <p>
+                                    Following show very simple example of the dictionary construct.
+                                    Normall, this will be within templates.xml file within addons.
```
```suggestion
                                    Normally, this will be within templates.xml file within addons.
```

elicoidal reviewed (approved):
Excellent job.
No test. Some light modifications but I did not check the technical side

bealdav commented on excel_import_export/wizard/import_xlsx_wizard.py:
```diff
@@ -0,0 +1,145 @@
+        readonly=True,
+        size=500,
+    )
+    datas = fields.Binary(
```
s/datas/data because data is plural form

kittiu replied:
This one I just replicate the file content field of ir_attachment.datas (I think it is naming convention, isn't it?)

bealdav replied:
non blocking on my side

bealdav reviewed (approved):
Really thanks for these features. I've installed and tested it. Very useful for functional users

Just a non blocking comment : maybe you may put in the roadmap to split in several modules when upgrade in upper odoo version because there is a lot of code here and it serves different features. It'll be easier to port by other people if features are split by modules.

IMO ready to merge except commit to clean and make travis works

Thanks a lot

kittiu commented:
@bealdav about roadmap. I think there is only 2 feature here, import and export. I am not sure I can make it more modular than this (both also share some function in common.py). But let's see, if it is what people need, we can do that.
My plan though, is to have following future roadmap.
- Adding button in xlsx.template, similar to server action, i.e.,  Add import action, Add export action. This can make it completely no coding.
- Module extension for Job Queue (connector), to allow running report by job (will be life saver for long running report).
- Anything else???

elicoidal commented:
I tend to agree with @kittiu it is difficult to cut the module in pieces.
About roadmap, LGTM (and nothing to add on my side for the moment :p)

kittiu commented:
It seem to have error using eval(), this method is not usable? What is the alternative.

![image](https://user-images.githubusercontent.com/1973598/53850877-4e020680-3fef-11e9-806e-b382c047941e.png)

moylop260 commented:
> It seem to have error using eval(), this method is not usable? What is the alternative.

[odoo.tools.safe_eval](https://github.com/odoo/odoo/blob/b5885f0cb02ba6853e17eb93009b1ebad2ff8d35/odoo/tools/safe_eval.py#L306)

moylop260 commented on excel_import_export/models/xlsx_export.py:
```diff
@@ -0,0 +1,271 @@
+                    value = eval(eval_cond, eval_context)
+                # Format w/Cond takes priority
+                format_cond = format_cond_dict[field[0]]
+                format = self._eval_format_cond(line._name, line,
```
`format` is a built-in method of python.
Then you are redefining it
What about using another variable name?

kittiu replied:
Changed all 'format' to 'style' (also Style is better meaning in our case too). Thanks.

kittiu committed:
Change from eval() to safe_evel()

kittiu committed:
Change variable to format to style, as fomat is a common python function

:100644 100644 00ee3d9f... e9e48d87... M	excel_import_export/models/common.py
:100644 100644 a215d29b... 5b4d1fb1... M	excel_import_export/models/styles.py
:100644 100644 ace11a32... 01e5b9f5... M	excel_import_export/models/xlsx_export.py
:100644 100644 881b814f... cadfb0f2... M	excel_import_export/models/xlsx_import.py
:100644 100644 58689ee5... 80490ce8... M	excel_import_export/models/xlsx_template.py
:100644 100644 5c9c09a6... a363ad19... M	excel_import_export/views/xlsx_template_view.xml
:100644 100644 475b5187... 392fe6e5... M	excel_import_export_demo/import_export_sale_order/templates.xml
:100644 100644 4af9c519... 45ee33c6... M	excel_import_export_demo/report_sale_order/templates.xml

:100644 100644 96157ea3... 17d3964d... M	excel_import_export/__manifest__.py
:100644 100644 00ee3d9f... 51c2572a... M	excel_import_export/models/common.py
:100644 100644 a215d29b... 5b4d1fb1... M	excel_import_export/models/styles.py
:100644 100644 ace11a32... 185a3330... M	excel_import_export/models/xlsx_export.py
:100644 100644 881b814f... cadfb0f2... M	excel_import_export/models/xlsx_import.py
:100644 100644 58689ee5... 80490ce8... M	excel_import_export/models/xlsx_template.py
:100644 100644 5c9c09a6... a363ad19... M	excel_import_export/views/xlsx_template_view.xml
:100644 100644 475b5187... 392fe6e5... M	excel_import_export_demo/import_export_sale_order/templates.xml
:100644 100644 4af9c519... 45ee33c6... M	excel_import_export_demo/report_sale_order/templates.xml

:100644 100644 96157ea3... 933ce0dc... M	excel_import_export/__manifest__.py
:100644 100644 00ee3d9f... 51c2572a... M	excel_import_export/models/common.py
:100644 100644 a215d29b... 5b4d1fb1... M	excel_import_export/models/styles.py
:100644 100644 ace11a32... 185a3330... M	excel_import_export/models/xlsx_export.py
:100644 100644 881b814f... cadfb0f2... M	excel_import_export/models/xlsx_import.py
:100644 100644 58689ee5... 80490ce8... M	excel_import_export/models/xlsx_template.py
:100644 100644 5c9c09a6... a363ad19... M	excel_import_export/views/xlsx_template_view.xml
:100644 100644 475b5187... 392fe6e5... M	excel_import_export_demo/import_export_sale_order/templates.xml
:100644 100644 4af9c519... 45ee33c6... M	excel_import_export_demo/report_sale_order/templates.xml

:100644 100644 96157ea3 3b1217e8 M	excel_import_export/__manifest__.py
:100644 100644 00ee3d9f 51c2572a M	excel_import_export/models/common.py
:100644 100644 a215d29b 5b4d1fb1 M	excel_import_export/models/styles.py
:100644 100644 ace11a32 185a3330 M	excel_import_export/models/xlsx_export.py
:100644 100644 881b814f cadfb0f2 M	excel_import_export/models/xlsx_import.py
:100644 100644 58689ee5 80490ce8 M	excel_import_export/models/xlsx_template.py
:100644 100644 5c9c09a6 a363ad19 M	excel_import_export/views/xlsx_template_view.xml
:100644 100644 475b5187 392fe6e5 M	excel_import_export_demo/import_export_sale_order/templates.xml
:100644 100644 4af9c519 45ee33c6 M	excel_import_export_demo/report_sale_order/templates.xml

:100644 100644 96157ea3 fee958bc M	excel_import_export/__manifest__.py
:100644 100644 00ee3d9f 51c2572a M	excel_import_export/models/common.py
:100644 100644 a215d29b 5b4d1fb1 M	excel_import_export/models/styles.py
:100644 100644 ace11a32 185a3330 M	excel_import_export/models/xlsx_export.py
:100644 100644 881b814f cadfb0f2 M	excel_import_export/models/xlsx_import.py
:100644 100644 58689ee5 80490ce8 M	excel_import_export/models/xlsx_template.py
:100644 100644 5c9c09a6 a363ad19 M	excel_import_export/views/xlsx_template_view.xml
:100644 100644 475b5187 392fe6e5 M	excel_import_export_demo/import_export_sale_order/templates.xml
:100644 100644 4af9c519 45ee33c6 M	excel_import_export_demo/report_sale_order/templates.xml

:100644 100644 96157ea3 fee958bc M	excel_import_export/__manifest__.py
:100644 100644 00ee3d9f 51c2572a M	excel_import_export/models/common.py
:100644 100644 a215d29b 9738a3c8 M	excel_import_export/models/styles.py
:100644 100644 ace11a32 a7d6adc5 M	excel_import_export/models/xlsx_export.py
:100644 100644 881b814f 12f9ca99 M	excel_import_export/models/xlsx_import.py
:100644 100644 70c37799 f123d2a6 M	excel_import_export/models/xlsx_report.py
:100644 100644 58689ee5 578a1fd8 M	excel_import_export/models/xlsx_template.py
:100644 100644 5c9c09a6 a363ad19 M	excel_import_export/views/xlsx_template_view.xml
:100644 100644 800ea573 1807ea7e M	excel_import_export/wizard/export_xlsx_wizard.py
:100644 100644 febed8d0 750dc17e M	excel_import_export/wizard/import_xlsx_wizard.py
:100644 100644 475b5187 392fe6e5 M	excel_import_export_demo/import_export_sale_order/templates.xml
:100644 100644 8e40a2d0 21574896 M	excel_import_export_demo/report_sale_order/report_sale_order.py
:100644 100644 4af9c519 45ee33c6 M	excel_import_export_demo/report_sale_order/templates.xml

:100644 100644 96157ea3 fee958bc M	excel_import_export/__manifest__.py
:100644 100644 00ee3d9f 51c2572a M	excel_import_export/models/common.py
:100644 100644 a215d29b 9738a3c8 M	excel_import_export/models/styles.py
:100644 100644 ace11a32 c7db3f92 M	excel_import_export/models/xlsx_export.py
:100644 100644 881b814f 12f9ca99 M	excel_import_export/models/xlsx_import.py
:100644 100644 70c37799 f123d2a6 M	excel_import_export/models/xlsx_report.py
:100644 100644 58689ee5 578a1fd8 M	excel_import_export/models/xlsx_template.py
:100644 100644 5c9c09a6 a363ad19 M	excel_import_export/views/xlsx_template_view.xml
:100644 100644 800ea573 1807ea7e M	excel_import_export/wizard/export_xlsx_wizard.py
:100644 100644 febed8d0 750dc17e M	excel_import_export/wizard/import_xlsx_wizard.py
:100644 100644 475b5187 392fe6e5 M	excel_import_export_demo/import_export_sale_order/templates.xml
:100644 100644 8e40a2d0 21574896 M	excel_import_export_demo/report_sale_order/report_sale_order.py
:100644 100644 4af9c519 45ee33c6 M	excel_import_export_demo/report_sale_order/templates.xml

:100644 100644 96157ea3 fee958bc M	excel_import_export/__manifest__.py
:100644 100644 00ee3d9f 51c2572a M	excel_import_export/models/common.py
:100644 100644 a215d29b 9738a3c8 M	excel_import_export/models/styles.py
:100644 100644 ace11a32 c7db3f92 M	excel_import_export/models/xlsx_export.py
:100644 100644 881b814f 12f9ca99 M	excel_import_export/models/xlsx_import.py
:100644 100644 70c37799 f123d2a6 M	excel_import_export/models/xlsx_report.py
:100644 100644 58689ee5 e3826e08 M	excel_import_export/models/xlsx_template.py
:000000 100644 00000000 34aa53bf A	excel_import_export/tests/__init__.py
:000000 100644 00000000 18618688 A	excel_import_export/tests/sale_order.xlsx
:000000 100644 00000000 c8481487 A	excel_import_export/tests/test_xlsx_template.py
:100644 100644 5c9c09a6 a363ad19 M	excel_import_export/views/xlsx_template_view.xml
:100644 100644 800ea573 1807ea7e M	excel_import_export/wizard/export_xlsx_wizard.py
:100644 100644 febed8d0 750dc17e M	excel_import_export/wizard/import_xlsx_wizard.py
:100644 100644 475b5187 392fe6e5 M	excel_import_export_demo/import_export_sale_order/templates.xml
:100644 100644 8e40a2d0 21574896 M	excel_import_export_demo/report_sale_order/report_sale_order.py
:100644 100644 4af9c519 45ee33c6 M	excel_import_export_demo/report_sale_order/templates.xml

:100644 100644 96157ea3 fee958bc M	excel_import_export/__manifest__.py
:100644 100644 00ee3d9f 51c2572a M	excel_import_export/models/common.py
:100644 100644 a215d29b 9738a3c8 M	excel_import_export/models/styles.py
:100644 100644 ace11a32 c7db3f92 M	excel_import_export/models/xlsx_export.py
:100644 100644 881b814f 12f9ca99 M	excel_import_export/models/xlsx_import.py
:100644 100644 70c37799 f123d2a6 M	excel_import_export/models/xlsx_report.py
:100644 100644 58689ee5 ed8c9fc7 M	excel_import_export/models/xlsx_template.py
:000000 100644 00000000 34aa53bf A	excel_import_export/tests/__init__.py
:000000 100644 00000000 18618688 A	excel_import_export/tests/sale_order.xlsx
:000000 100644 00000000 69aa6ea0 A	excel_import_export/tests/test_xlsx_template.py
:100644 100644 5c9c09a6 a363ad19 M	excel_import_export/views/xlsx_template_view.xml
:100644 100644 800ea573 1807ea7e M	excel_import_export/wizard/export_xlsx_wizard.py
:100644 100644 febed8d0 750dc17e M	excel_import_export/wizard/import_xlsx_wizard.py
:100644 100644 475b5187 392fe6e5 M	excel_import_export_demo/import_export_sale_order/templates.xml
:100644 100644 8e40a2d0 21574896 M	excel_import_export_demo/report_sale_order/report_sale_order.py
:100644 100644 4af9c519 45ee33c6 M	excel_import_export_demo/report_sale_order/templates.xml

:100644 100644 96157ea3 fee958bc M	excel_import_export/__manifest__.py
:100644 100644 00ee3d9f 51c2572a M	excel_import_export/models/common.py
:100644 100644 a215d29b 9738a3c8 M	excel_import_export/models/styles.py
:100644 100644 ace11a32 c7db3f92 M	excel_import_export/models/xlsx_export.py
:100644 100644 881b814f 933d8614 M	excel_import_export/models/xlsx_import.py
:100644 100644 70c37799 f123d2a6 M	excel_import_export/models/xlsx_report.py
:100644 100644 58689ee5 1460473a M	excel_import_export/models/xlsx_template.py
:100644 100644 5c9c09a6 a363ad19 M	excel_import_export/views/xlsx_template_view.xml
:100644 100644 800ea573 1807ea7e M	excel_import_export/wizard/export_xlsx_wizard.py
:100644 100644 febed8d0 750dc17e M	excel_import_export/wizard/import_xlsx_wizard.py
:100644 100644 a2d035ef 9463f279 M	excel_import_export_demo/__manifest__.py
:100644 100644 475b5187 e7f1255b M	excel_import_export_demo/import_export_sale_order/templates.xml
:100644 100644 8e40a2d0 21574896 M	excel_import_export_demo/report_sale_order/report_sale_order.py
:100644 100644 4af9c519 45ee33c6 M	excel_import_export_demo/report_sale_order/templates.xml
:000000 100644 00000000 79db62f7 A	excel_import_export_demo/tests/__init__.py
:000000 100644 00000000 18618688 A	excel_import_export_demo/tests/sale_order.xlsx
:000000 100644 00000000 c9733b95 A	excel_import_export_demo/tests/test_common.py
:000000 100644 00000000 9c943768 A	excel_import_export_demo/tests/test_xlsx_import_export.py
:000000 100644 00000000 730605c1 A	excel_import_export_demo/tests/test_xlsx_template.py

:100644 100644 96157ea3 fee958bc M	excel_import_export/__manifest__.py
:100644 100644 00ee3d9f 51c2572a M	excel_import_export/models/common.py
:100644 100644 a215d29b 9738a3c8 M	excel_import_export/models/styles.py
:100644 100644 ace11a32 c7db3f92 M	excel_import_export/models/xlsx_export.py
:100644 100644 881b814f 933d8614 M	excel_import_export/models/xlsx_import.py
:100644 100644 70c37799 f123d2a6 M	excel_import_export/models/xlsx_report.py
:100644 100644 58689ee5 1460473a M	excel_import_export/models/xlsx_template.py
:100644 100644 5c9c09a6 a363ad19 M	excel_import_export/views/xlsx_template_view.xml
:100644 100644 800ea573 1807ea7e M	excel_import_export/wizard/export_xlsx_wizard.py
:100644 100644 febed8d0 750dc17e M	excel_import_export/wizard/import_xlsx_wizard.py
:100644 100644 a2d035ef 9463f279 M	excel_import_export_demo/__manifest__.py
:100644 100644 475b5187 e7f1255b M	excel_import_export_demo/import_export_sale_order/templates.xml
:100644 100644 8e40a2d0 21574896 M	excel_import_export_demo/report_sale_order/report_sale_order.py
:100644 100644 4af9c519 45ee33c6 M	excel_import_export_demo/report_sale_order/templates.xml
:000000 100644 00000000 79db62f7 A	excel_import_export_demo/tests/__init__.py
:000000 100644 00000000 18618688 A	excel_import_export_demo/tests/sale_order.xlsx
:000000 100644 00000000 bb3ea32e A	excel_import_export_demo/tests/test_common.py
:000000 100644 00000000 9c943768 A	excel_import_export_demo/tests/test_xlsx_import_export.py
:000000 100644 00000000 730605c1 A	excel_import_export_demo/tests/test_xlsx_template.py

kittiu commented:
> > It seem to have error using eval(), this method is not usable? What is the alternative.
> 
> [odoo.tools.safe_eval](https://github.com/odoo/odoo/blob/b5885f0cb02ba6853e17eb93009b1ebad2ff8d35/odoo/tools/safe_eval.py#L306)

Done, thanks.

kittiu commented:
Hi. I can make it green for travis. But on runbot I dont see where it look at error. I click on build log and it shows 404 not found error. Any hint?
Many thanks.

elicoidal commented:
I do have access to runbot:
![image](https://user-images.githubusercontent.com/7600613/53928613-bde2c080-40c5-11e9-885a-b260d3999974.png)

kittiu commented:
> I do have access to runbot:
> ![image](https://user-images.githubusercontent.com/7600613/53928613-bde2c080-40c5-11e9-885a-b260d3999974.png)

Yes, but can't get the log.
![00422](https://user-images.githubusercontent.com/1973598/53930097-9d653700-40c2-11e9-859e-fa7f6e833d20.png)

elicoidal commented:
I can click on full logs and get some info (not sure that is enough though)

@moylop260 @gurneyalex when clicking on build logs there is a 404. Can you have a look?

kittiu commented:
Hi there. Runbot is green now.
Now, we have another 2 codecov came up. How to pass it?

yajo commented:
To pass those tests, you need to add unit tests to your module. Check this:
![peek 08-03-2019 09-14](https://user-images.githubusercontent.com/973709/54019243-9bfb4380-4182-11e9-9be9-a62d46d1893c.gif)

The lines you see in red are never executed. If all lines were green, you'd be getting 100% coverage. You should write unit tests that somehow execute those lines by, i.e., calling the methods and checking their outputs.

Check [the docs on how to write unit tests](https://www.odoo.com/documentation/12.0/reference/testing.html).

kittiu commented:
Hi all, I am in the process of writing test. Is it possible that, I run test case in excel_import_export_demo.
Because to run test I need to test with document, i.e., sale.order, which is only available when excel_import_export_demo is install.

Note: With excel_import_export alone, which is only depends on 'mail', there is no model that make sense (model with header doc, doc's lines) to write test against.

bealdav commented:
Yes it's pragmatic to put tests in excel_import_export_demo

The important thing is to have tests to secure module.

kittiu commented:
> Yes it's pragmatic to put tests in excel_import_export_demo
> 
> The important thing is to have tests to secure module.

Great ! Thanks !

bealdav commented:
Really thanks for this clever implementation of this use case

kittiu commented:
Yeah! Thank you all for your advices.

kittiu commented:
After 2 weeks, I still can't find this addon in Odoo App Store. What is the criteria to be there?

elicoidal commented:
it is available in our apps platform :tongue: https://odoo-community.org/shop?&search=excel_import

elicoidal commented:
Not sure why not in the other one: the scan is at repo level

kittiu commented:
Thank you, so I assume to just wait. :+1:

elicoidal commented:
Not exactly: it should be scanned every day so this is not normal: we will check the apps settings and contact Odoo.
@pedrobaeza are you the right person to do that?
